File: backend/app/core/config_loader.py
import json
import os
from pathlib import Path
from typing import Dict, Any, List, Optional
import threading
import time

from app.core.logger import get_logger

logger = get_logger(__name__)


class ConfigLoader:
    _instance = None
    _lock = threading.Lock()
    
    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if self._initialized:
            return
        self._initialized = True
        
        self.base_path = Path(__file__).parent.parent.parent / "config"
        self._config_cache: Dict[str, Any] = {}
        self._last_modified: Dict[str, float] = {}
        self._current_data_source_type = 'file'
        self._db_session_factory = None
        self._data_source = None
        
        self._load_app_config()
        self._load_recommendations()
        self._load_prompts()
        self._load_ontologies()
        self._load_scenes()
    
    def get_current_data_source_type(self) -> str:
        return self._current_data_source_type
    
    def set_db_session_factory(self, session_factory):
        self._db_session_factory = session_factory
    
    def switch_data_source(self, source_type):
        from app.core.data_source import DataSourceType
        
        if source_type == DataSourceType.DATABASE:
            if not self._db_session_factory:
                raise RuntimeError("[ConfigLoader] 切换到数据库数据源前，必须先调用 set_db_session_factory 设置会话工厂")
            
            from app.core.database_data_source import DatabaseDataSource
            self._data_source = DatabaseDataSource(self._db_session_factory)
            
            self._config_cache['ontologies'] = self._data_source.load_ontologies()
            self._config_cache['scenes'] = self._data_source.load_scenes()
            self._config_cache['prompts'] = self._data_source.load_prompts()
            self._config_cache['recommendations'] = self._data_source.load_recommendations()
            
            self._current_data_source_type = 'database'
            logger.info("[ConfigLoader] 数据源已切换为 database")
        
        elif source_type == DataSourceType.FILE:
            self._data_source = None
            self._load_app_config()
            self._load_recommendations()
            self._load_prompts()
            self._load_ontologies()
            self._load_scenes()
            self._current_data_source_type = 'file'
            logger.info("[ConfigLoader] 数据源已切换为 file")
        
        else:
            raise ValueError(f"[ConfigLoader] 不支持的数据源类型: {source_type}")
    
    def _load_json(self, path: Path) -> Optional[Dict]:
        try:
            with open(path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            logger.debug("[ConfigLoader] 加载 JSON: %s", path.name)
            return data
        except Exception as e:
            logger.error("[ConfigLoader] 加载失败 %s: %s", path, e)
            return None
    
    def _load_text(self, path: Path) -> Optional[str]:
        try:
            with open(path, 'r', encoding='utf-8') as f:
                data = f.read()
            logger.debug("[ConfigLoader] 加载 TXT: %s (%d 字符)", path.name, len(data))
            return data
        except Exception as e:
            logger.error("[ConfigLoader] 加载失败 %s: %s", path, e)
            return None
    
    def _load_app_config(self):
        path = self.base_path / "app_config.json"
        if path.exists():
            app_config = self._load_json(path)
            self._config_cache['app_config'] = app_config
            self._last_modified['app_config'] = path.stat().st_mtime
            
            if app_config:
                system_config = {}
                for key in ['recommendation', 'smartRecommend', 'sceneRecognition', 'fieldExtraction']:
                    if key in app_config:
                        system_config[key] = app_config[key]
                self._config_cache['system_config'] = system_config
    
    def _load_recommendations(self):
        path = self.base_path / "templates" / "recommendations.json"
        if path.exists():
            data = self._load_json(path)
            if data:
                self._config_cache['recommendations'] = data.get('recommendations', {})
                self._last_modified['recommendations'] = path.stat().st_mtime
    
    def _load_prompts(self):
        prompts = {}
        prompts_path = self.base_path / "prompts"
        if prompts_path.exists():
            for file in prompts_path.glob("*.txt"):
                prompt_name = file.stem
                prompts[prompt_name] = self._load_text(file)
        
        scene_prompts = {}
        scene_prompts_path = self.base_path / "prompts" / "scenes"
        if scene_prompts_path.exists():
            for file in scene_prompts_path.glob("*.txt"):
                prompt_name = file.stem
                scene_prompts[prompt_name] = self._load_text(file)
        
        self._config_cache['prompts'] = prompts
        self._config_cache['scene_prompts'] = scene_prompts
    
    def _load_ontologies(self):
        ontologies = {}
        ontologies_path = self.base_path / "ontologies"
        if ontologies_path.exists():
            for file in ontologies_path.glob("*.json"):
                data = self._load_json(file)
                if data:
                    ontology_code = file.stem
                    ontologies[ontology_code] = data
                    logger.debug("[ConfigLoader] 加载本体: %s", ontology_code)
        
        self._config_cache['ontologies'] = ontologies
        logger.info("[ConfigLoader] 从文件加载本体 count=%d", len(ontologies))
    
    def _load_scenes(self):
        scenes = []
        scene_mapping_path = self.base_path / "scenes" / "scene_mapping.json"
        if scene_mapping_path.exists():
            data = self._load_json(scene_mapping_path)
            if data:
                scenes = data.get("sceneMappings", [])
        
        versions_path = self.base_path / "versions"
        if versions_path.exists():
            for scene_dir in versions_path.iterdir():
                if scene_dir.is_dir():
                    for file in scene_dir.glob("*.json"):
                        data = self._load_json(file)
                        if data:
                            scene_data = data.get("data", data)
                            scene_code = scene_data.get("sceneCode", scene_dir.name)
                            scenes.append({
                                "sceneCode": scene_code,
                                "sceneName": scene_data.get("sceneName", scene_code),
                                "description": scene_data.get("description", ""),
                                "keywords": scene_data.get("keywords", []),
                                "priority": scene_data.get("priority", 1),
                                "isActive": scene_data.get("isActive", True),
                                "promptCode": scene_data.get("promptCode", scene_code),
                                "actionPrompt": scene_code,
                                "config": scene_data.get("config", {})
                            })
        
        self._config_cache['scenes'] = scenes
        logger.info("[ConfigLoader] 从文件加载场景 count=%d", len(scenes))
    
    def get_app_config(self) -> Dict[str, Any]:
        return self._config_cache.get('app_config', {})
    
    def get_scene_mappings(self) -> List[Dict]:
        return self._config_cache.get('scenes', [])
    
    def get_scene_by_code(self, scene_code: str) -> Optional[Dict]:
        scenes = self._config_cache.get('scenes', [])
        for scene in scenes:
            if scene.get('sceneCode') == scene_code:
                return scene
        return None
    
    def get_all_scenes(self) -> List[Dict]:
        return self.get_scene_mappings()
    
    def get_scene_prompt(self, scene_code: str) -> Optional[str]:
        scene = self.get_scene_by_code(scene_code)
        if scene:
            prompt_file = scene.get('actionPrompt')
            if prompt_file:
                scene_prompts = self._config_cache.get('scene_prompts', {})
                prompt_name = Path(prompt_file).stem
                return scene_prompts.get(prompt_name)
        return None
    
    def get_ontology(self, form_code: str) -> Optional[Dict]:
        return self._config_cache.get('ontologies', {}).get(form_code)
    
    def get_all_ontologies(self) -> Dict[str, Dict]:
        return self._config_cache.get('ontologies', {})
    
    def get_recommendations(self, form_code: str, field_code: str) -> List[str]:
        recommendations = self._config_cache.get('recommendations', {})
        form_recommendations = recommendations.get(form_code, {})
        return form_recommendations.get(field_code, [])
    
    def get_prompt(self, prompt_name: str) -> Optional[str]:
        prompts = self._config_cache.get('prompts', {})
        if prompt_name in prompts:
            return prompts[prompt_name]
        scene_prompts = self._config_cache.get('scene_prompts', {})
        return scene_prompts.get(prompt_name)
    
    def get_system_config(self) -> Dict[str, Any]:
        return self._config_cache.get('system_config', {})
    
    def get_recommendation_config(self) -> Dict[str, Any]:
        return self._config_cache.get('system_config', {}).get('recommendation', {})
    
    def get_scene_recognition_config(self) -> Dict[str, Any]:
        return self._config_cache.get('system_config', {}).get('sceneRecognition', {})
    
    def get_field_extraction_config(self) -> Dict[str, Any]:
        return self._config_cache.get('system_config', {}).get('fieldExtraction', {})
    
    def reload_config(self, config_type: Optional[str] = None):
        if config_type is None or config_type == 'all':
            self._load_app_config()
            self._load_recommendations()
            self._load_prompts()
            self._load_ontologies()
            self._load_scenes()
        elif config_type == 'system_config':
            self._load_app_config()
        elif config_type == 'app_config':
            self._load_app_config()
        elif config_type == 'ontologies':
            self._load_ontologies()
        elif config_type == 'recommendations':
            self._load_recommendations()
        elif config_type == 'prompts':
            self._load_prompts()
        elif config_type == 'scenes':
            self._load_scenes()


config_loader = ConfigLoader()

File: backend-app/src/main/java/com/sitech/prodai/config/ConfigLoader.java
package com.sitech.prodai.config;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PostConstruct;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import org.springframework.stereotype.Component;

import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * 配置加载器 —— 对齐 Python {@code app/core/config_loader.py::ConfigLoader}。
 *
 * <p>从 classpath 加载本体 / 场景 / 提示词 / 应用配置 / 推荐模板，
 * 对应 Python 从 {@code backend/config/} 目录加载的文件数据源。
 *
 * <p>Python 资源目录 → Java classpath 映射：
 * <ul>
 *   <li>backend/config/ontologies/*.json        → classpath:ontologies/*.json</li>
 *   <li>backend/config/scenes/scene_mapping.json→ classpath:scenes/scene_mapping.json</li>
 *   <li>backend/config/prompts/*.txt            → classpath:prompts/*.txt</li>
 *   <li>backend/config/prompts/scenes/*.txt     → classpath:prompts/scenes/*.txt</li>
 *   <li>backend/config/app_config.json          → classpath:app_config.json</li>
 *   <li>backend/config/templates/recommendations.json → classpath:templates/recommendations.json</li>
 * </ul>
 */
@Component
public class ConfigLoader {

    private static final Logger log = LoggerFactory.getLogger(ConfigLoader.class);

    private final ObjectMapper objectMapper;
    private final PathMatchingResourcePatternResolver resolver = new PathMatchingResourcePatternResolver();

    /** 本体定义：ontologyCode → 原始 JSON */
    private final Map<String, Map<String, Object>> ontologies = new ConcurrentHashMap<>();
    /** 场景列表 */
    private final List<Map<String, Object>> scenes = new ArrayList<>();
    /** 通用提示词：promptName → 文本 */
    private final Map<String, String> prompts = new ConcurrentHashMap<>();
    /** 场景提示词：promptName → 文本 */
    private final Map<String, String> scenePrompts = new ConcurrentHashMap<>();
    /** 应用配置（app_config.json 全量） */
    private final Map<String, Object> appConfig = new ConcurrentHashMap<>();
    /** 系统配置（app_config 中提取的子集） */
    private final Map<String, Object> systemConfig = new ConcurrentHashMap<>();
    /** 推荐模板：formCode → {fieldCode → [values]} */
    private final Map<String, Map<String, Object>> recommendations = new ConcurrentHashMap<>();

    public ConfigLoader(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    @PostConstruct
    public void init() {
        reloadAll();
    }

    /** 全量重新加载（对齐 Python reload_config(None)） */
    public synchronized void reloadAll() {
        loadAppConfig();
        loadRecommendations();
        loadPrompts();
        loadOntologies();
        loadScenes();
        log.info("[ConfigLoader] 配置加载完成 ontologies={}, scenes={}, prompts={}, scenePrompts={}",
                ontologies.size(), scenes.size(), prompts.size(), scenePrompts.size());
    }

    // ==================== 加载方法 ====================

    private void loadAppConfig() {
        appConfig.clear();
        systemConfig.clear();
        try {
            Resource resource = resolver.getResource("classpath:app_config.json");
            if (resource.exists()) {
                try (InputStream in = resource.getInputStream()) {
                    Map<String, Object> raw = objectMapper.readValue(in, new TypeReference<>() {
                    });
                    appConfig.putAll(raw);
                    // 提取系统配置子集（对齐 Python _load_app_config）
                    for (String key : new String[]{"recommendation", "smartRecommend", "sceneRecognition", "fieldExtraction"}) {
                        if (raw.containsKey(key)) {
                            systemConfig.put(key, raw.get(key));
                        }
                    }
                }
            }
        } catch (Exception e) {
            log.warn("[ConfigLoader] 加载 app_config.json 失败: {}", e.getMessage());
        }
    }

    private void loadRecommendations() {
        recommendations.clear();
        try {
            Resource resource = resolver.getResource("classpath:templates/recommendations.json");
            if (resource.exists()) {
                try (InputStream in = resource.getInputStream()) {
                    Map<String, Object> raw = objectMapper.readValue(in, new TypeReference<>() {
                    });
                    Object recs = raw.get("recommendations");
                    if (recs instanceof Map<?, ?> map) {
                        for (Map.Entry<?, ?> e : map.entrySet()) {
                            if (e.getKey() instanceof String k && e.getValue() instanceof Map<?, ?> v) {
                                @SuppressWarnings("unchecked")
                                Map<String, Object> cast = (Map<String, Object>) v;
                                recommendations.put(k, cast);
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            log.warn("[ConfigLoader] 加载 recommendations.json 失败: {}", e.getMessage());
        }
    }

    private void loadPrompts() {
        prompts.clear();
        scenePrompts.clear();
        loadTextDir("classpath*:prompts/*.txt", prompts);
        loadTextDir("classpath*:prompts/scenes/*.txt", scenePrompts);
    }

    private void loadTextDir(String pattern, Map<String, String> target) {
        try {
            Resource[] resources = resolver.getResources(pattern);
            for (Resource resource : resources) {
                String filename = resource.getFilename();
                if (filename == null || !filename.endsWith(".txt")) {
                    continue;
                }
                String name = filename.substring(0, filename.length() - 4);
                try (InputStream in = resource.getInputStream()) {
                    String text = new String(in.readAllBytes(), StandardCharsets.UTF_8);
                    target.put(name, text);
                }
            }
        } catch (Exception e) {
            log.warn("[ConfigLoader] 加载提示词目录失败 pattern={} err={}", pattern, e.getMessage());
        }
    }

    private void loadOntologies() {
        ontologies.clear();
        try {
            Resource[] resources = resolver.getResources("classpath*:ontologies/*.json");
            for (Resource resource : resources) {
                String filename = resource.getFilename();
                if (filename == null) {
                    continue;
                }
                String code = filename.substring(0, filename.length() - 5);
                try (InputStream in = resource.getInputStream()) {
                    Map<String, Object> data = objectMapper.readValue(in, new TypeReference<>() {
                    });
                    ontologies.put(code, data);
                }
            }
        } catch (Exception e) {
            log.warn("[ConfigLoader] 加载本体失败: {}", e.getMessage());
        }
        log.info("[ConfigLoader] 从文件加载本体 count={}", ontologies.size());
    }

    @SuppressWarnings("unchecked")
    private void loadScenes() {
        scenes.clear();
        try {
            Resource resource = resolver.getResource("classpath:scenes/scene_mapping.json");
            if (resource.exists()) {
                try (InputStream in = resource.getInputStream()) {
                    Map<String, Object> data = objectMapper.readValue(in, new TypeReference<>() {
                    });
                    Object mappings = data.get("sceneMappings");
                    if (mappings instanceof List<?> list) {
                        for (Object item : list) {
                            if (item instanceof Map<?, ?> m) {
                                scenes.add((Map<String, Object>) m);
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            log.warn("[ConfigLoader] 加载场景映射失败: {}", e.getMessage());
        }
        log.info("[ConfigLoader] 从文件加载场景 count={}", scenes.size());
    }

    // ==================== 查询方法（对齐 Python ConfigLoader） ====================

    public Map<String, Object> getAppConfig() {
        return appConfig;
    }

    public List<Map<String, Object>> getSceneMappings() {
        return scenes;
    }

    public Map<String, Object> getSceneByCode(String sceneCode) {
        for (Map<String, Object> scene : scenes) {
            if (sceneCode.equals(scene.get("sceneCode"))) {
                return scene;
            }
        }
        return null;
    }

    public List<Map<String, Object>> getAllScenes() {
        return scenes;
    }

    public String getScenePrompt(String sceneCode) {
        Map<String, Object> scene = getSceneByCode(sceneCode);
        if (scene == null) {
            return null;
        }
        Object actionPrompt = scene.get("actionPrompt");
        if (actionPrompt == null) {
            actionPrompt = scene.get("promptCode");
        }
        if (actionPrompt == null) {
            return null;
        }
        String promptName = String.valueOf(actionPrompt);
        int dot = promptName.indexOf('.');
        if (dot > 0) {
            promptName = promptName.substring(0, dot);
        }
        return scenePrompts.getOrDefault(promptName, prompts.get(promptName));
    }

    public Map<String, Object> getOntology(String formCode) {
        return ontologies.get(formCode);
    }

    public Map<String, Map<String, Object>> getAllOntologies() {
        return ontologies;
    }

    @SuppressWarnings("unchecked")
    public List<String> getRecommendations(String formCode, String fieldCode) {
        Map<String, Object> formRecs = recommendations.get(formCode);
        if (formRecs == null) {
            return Collections.emptyList();
        }
        Object values = formRecs.get(fieldCode);
        if (values instanceof List<?> list) {
            List<String> result = new ArrayList<>();
            for (Object v : list) {
                result.add(v == null ? null : String.valueOf(v));
            }
            return result;
        }
        return Collections.emptyList();
    }

    public String getPrompt(String promptName) {
        if (prompts.containsKey(promptName)) {
            return prompts.get(promptName);
        }
        return scenePrompts.get(promptName);
    }

    public Map<String, Object> getSystemConfig() {
        return systemConfig;
    }

    public Map<String, Object> getRecommendationConfig() {
        Object rec = systemConfig.get("recommendation");
        if (rec instanceof Map<?, ?> map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> cast = (Map<String, Object>) map;
            return cast;
        }
        return Collections.emptyMap();
    }

    public Map<String, Object> getSceneRecognitionConfig() {
        Object cfg = systemConfig.get("sceneRecognition");
        if (cfg instanceof Map<?, ?> map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> cast = (Map<String, Object>) map;
            return cast;
        }
        return Collections.emptyMap();
    }

    public Map<String, Object> getFieldExtractionConfig() {
        Object cfg = systemConfig.get("fieldExtraction");
        if (cfg instanceof Map<?, ?> map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> cast = (Map<String, Object>) map;
            return cast;
        }
        return Collections.emptyMap();
    }

    public void reloadConfig(String configType) {
        if (configType == null || "all".equals(configType)) {
            reloadAll();
        } else {
            switch (configType) {
                case "system_config", "app_config" -> loadAppConfig();
                case "ontologies" -> loadOntologies();
                case "recommendations" -> loadRecommendations();
                case "prompts" -> loadPrompts();
                case "scenes" -> loadScenes();
                default -> log.warn("[ConfigLoader] 未知配置类型: {}", configType);
            }
        }
    }

    /** 占位：数据库数据源切换（对齐 Python switch_data_source），后续阶段实现 */
    public String getCurrentDataSourceType() {
        return "file";
    }
}
